Add Changelog file for continuous updating

Kebap opened the pull request:
<!-- Keep the title short & concise so your mom can understand it -->
#### Brief overview of PR changes/additions

Add CHANGELOG.md file with historic info, which should be updated each commit way before release time, see discussion in #3156

#### Motivation for adding to Mudlet

- Make it easier for everyone to grasp recent changes quickly.
- Continuously collect information needed to prepare releases.


#### Other info (issues closed, discussion etc)
* [x] Before merge, need to remove commit of  .github/workflows/update-translations.yml still, but codespaces was not available

Fix #3156

While #4818 added a section to PR text, this does not go far enough, does not ask about which section the brief summary is intended for, does not allow opportunity to sort and style entries, does not show recent changelog for comparison, etc.

Doubts were mentioned about the burden added for first time contributors. The most commits come from core team though. We try and establish common best practices. After that, include directions in CONTRIBUTING.md file. We always go easy on first time contributors anyway. If no simple brief line can be added to changelog, that commit needs to be handled manually, just like today ALL commits must be checked before release time. This change should reduce that work to a bare minimum.

Maybe compare how other projects handle their changelogs in comparison, see what we can learn, e.g. translating it in Crowdin as well: https://github.com/cgeo/cgeo/issues/4317

#### Release post highlight
infrastructure: Added changelog file to github, continuously describing changes between releases

Kebap committed:
Update update-translations.yml

Kebap committed:
Create a changelog

Kebap committed:
Merge branch 'development' into changelog

Kebap committed:
Add upcoming versions

Kebap committed:
Highlight unreleased state

Kebap commented:
Not sure how much I like the spoilers really. I may remove them again.
They make it easy to find the correct release version, condensing total text shown nicely.
Then again, they make it hard to link to individual paragraphs hidden inside, and scroll around for impressions.

Kebap committed:
Remove spoilers [skip ci]

SlySven commented:
I would note that in some environments a CHANGELOG is strongly desired (Debian) if not mandated (GNU native projects)!

demonnic commented:
I'm all for a proper changelog.

If we're gonna add a file, why not establish sections/conventions for items moving forward now, rather than waiting until it gets out of hand/folks develop bad habits? 

I'd suggest perhaps the following:

* Breaking Changes
  * even though we rarely make such, even more reason to call them out first
  * we should still include it, just empty, when there are no breaking changes.
  * can make it a fun statistic too... "releases/months since last breaking change"
* Enhancements
  * new stuff, additional features, etc
* Bugfixes
  * what it says on the tin
* Infrastructure
  * github actions, CI, stuff like that
  * list it last cuz it's mostly only stuff the core team/nerds will care about

If we do it with the top section being 'PTB' for all changes in the upcoming release, and released versions listed beneath it, then the release notes could fairly easily be automated for release posts. 

Slightly off-topic but related, it would also make it fairly easy to go through and bulk replace 'PTB' in the wiki as a mudlet version in the api docs to the version just released. This allows us to clearly mark functions which are only in the PTB without adding a lot of fiddly manual overhead. 

Especially if we take the time now to actually require things like changelog entries in order for PRs to pass. I suggest something like https://danger.systems/ruby/ for this, as I have used this in the past and use it currently at one of my places of employment, and it makes things like this incredibly easy to automate. We could add a label for automated PRs for things like translation updates and autocomplete scrapes which caused it to bypass the check. I'm willing to work on setting this aspect of things up, especially as I've already got it setup elsewhere and would just need to make it a github action which could be run and tweak things for our particulars.

If the concern is strain on first time contributors, I don't think requiring at least one line in the changelog is a lot, but a member of the core team can always step in and help them out by making the adjustment for them as a suggestion in the PR itself, which should alleviate that.

As the project continues to grow and developers add their efforts it only becomes more important to craft a set of standards and adhere to them. Until it's decided a standard needs to change, and then you change it. 

All of which is to say I agree with this, but think we should honestly go even further if we're going to go this far.

Edru2 commented:
I think I myself am pretty bad at writing this kind of things but I think the overall idea is good even though (as discussed already) the quality will differ. 
Where exactly will this changelog be visible?

Delwing commented:
I am throwing here my thoughts. 

- this file will be major source of git conflicts, therefore with each merge conflicts will likely need to be resolved (even though usually it will be easy conflict - accepting both options)
- this of course eases burden of releases and preparing and refining change log
- changelog entries are tied to code, even can be checked in CI flows whether PR is providing changelog modification
- there is no easy way to redact changes, you can only commit or suggest changes
- some people have better skills in writing "prose", some worse :)

Generally idea is good, not sure about form though. There are definitely some pros and cons here to consider.

SlySven commented:
> * this file will be major source of git conflicts, therefore with each merge conflicts will likely need to be resolved (even though usually it will be easy conflict - accepting both options)

How many lines does Git consider when working out which lines conflict - if we always add new lines at the bottom or top of a section will leaving one blank line between them be enough to reduce the conflicts - or should we sort by first word(s) in the texts or what?

demonnic commented:
Git is usually pretty good about resolving minor issues like this, in my experience, and the few times the changelog has been an issue it was easy enough to resolve using the web interface.

Delwing commented:
If same line is modified in different branches they will conflict upon merge.
Git is pretty good in resolving conflicts, yet it will be tedious to resolve them often.
If we want to avoid conflicts instead of creating single file for each branch in changelog DIRECTORY. So if we name files after branch, we will never have conflicts. Upon release all files should be merged automatically into single changelog nad removed.
As a release "partials" we could either use markdown or something more structured like `.json` or .`yml`

Delwing commented:
As one could suspect,  we're not first ones to face the problem.

https://about.gitlab.com/blog/2018/07/03/solving-gitlabs-changelog-conflict-crisis/
https://github.com/KiloKilo/changelog
https://pspdfkit.com/blog/2018/the-challenges-of-changelogs/

And even more obvious, yet harder to enforce, I doubt that this one can work outside of very small disciplined team (since commit message would be used directly)
https://upte.ch/blog/keep-a-changelog-without-conflicts/

SlySven commented on CHANGELOG.md:
```diff
@@ -0,0 +1,491 @@
+
+## macOS 10.13 High Sierra
+
+Apple will stop supporting 10.13 High Sierra in November 2020, and Mudlet will have to follow suit. Thus Mudlet 4.10 will be the last version to support 10.13, which had a good run of 3 years!
```
Needs edit now doesn't it?
```suggestion
Apple has stopped supporting 10.13 High Sierra since November 2020, and Mudlet had to follow suit. Thus Mudlet 4.10 was the last version to support 10.13, which had a good run of 3 years!
```

Kebap replied:
This was a verbatim copy of the website release news post.
We never go back to update those at all, at all.
Isn't it funny how easy this now becomes?
For CHANGELOG, keeping it current may make sense indeed.

SlySven commented on CHANGELOG.md:
```diff
@@ -0,0 +1,491 @@
+- option not to highlight text (history) when pressing the ↑ button
+- option to use your own pathfinding algorithm instead of the built-in one
+- you can now create standalone commandlines now allow you to ask the player for info
+- transparency in now supported by cecho/decho/hecho
```
If we are sorting these alphabetically then this needs to move up one line!

Kebap replied:
I don't think we are going to though

wiploo commented:
Mhhh, I don't think that a changelog.md is needed (I love changelos :-D)

When mudlet's releases are monthly based not many changes are listend in news page (10-20 rows). Now there is a big changelog because area months that mudlet isn't release, but it isn't the normal working model.

Why not create a changelog only upon releases? Avoiding to get mad making them over-and-over on every PR?

vadi2 commented:
I'm skeptical of the idea as well because of the burden it'll place on coders - asking one person to wear the hat of a coder, a documentation writer, and a copywriter is a fair bit. I realise that's ironic as I'm the one who does the release posts. This does have some level of support so let's try this for a couple of months to see how people feel about it then.

demonnic commented:
Y'all do what you want. Just let me know what you decide.

Kebap commented:
Thanks for all your thoughts and feedback, especially doubts and questions, but also support and additions.
I will try comment on each and add my thoughts in between to hopefully settle some issues.
This week the release of 4.11 proper will block implementing this merge anyway. 

@SlySven
> in some environments a CHANGELOG is strongly desired (Debian) if not mandated (GNU native projects)!

That's one of the benefits already discussed in the original suggestion.
Also updating changelog each pr makes it easier to test ptb & rc versions.
Then again, it may also lift some burdens off of the release post writers.

On the other hand, we need to carefully weigh and balance the downsides.
Some new contributors may be overwhelmed by a new requirement like this.
Others may not want to or simply fail to describe the changes concicely.

In those cases, we can hint text or even submit a PR to their PR with it.
All of this and more is indeed discussed in #3156 - see there for details.

@demonnic
> If we're gonna add a file, why not establish sections/conventions for items moving forward now, rather than waiting until it gets out of hand/folks develop bad habits?

The contents in CHANGELOG here are verbatim copies of the existing release posts.
If you scroll (or rather click) through those, you'll find conventions developed already.
However they have not been set in stone so far. This can surely become a matter of taste.

> 1. Breaking Changes  
>    * even though we rarely make such, even more reason to call them out first
>    * we should still include it, just empty, when there are no breaking changes.
>    * can make it a fun statistic too... "releases/months since last breaking change"
> 2. Enhancements
>    * new stuff, additional features, etc
> 3. Bugfixes 
> 4. Infrastructure
>    * github actions, CI, stuff like that
>    * list it last cuz it's mostly only stuff the core team/nerds will care about

Actually before items' sections we do a few "spot lights" on outstanding improvements.
There can be ample discussion about the "best" structure for items' sections really.
We currently use mostly the following sections for one-liners in that very order:

> 5. Added
> 6. Improved
> 7. Fixed
> 8. Infrastructure

Comparing these then, I find your request mostly fulfilled already.
Where sections 7 and 8 seem to mirror the suggested 3 and 4 closely.
Section 2. has been split into 5 and 6 like you commented as well.

The section 1 is not reflected in template. We don't want to encourage those at all.
Agree, if any were to happen, they should be mentioned at top or in spot lights even.
Not sure how fun "we did not break anything" stays, especially after regular hotfixes.

Alternatively, these are the sections recommended by keepachangelog.com 

> 9. Added - for new features.
> 10. Changed - for changes in existing functionality.
> 11. Deprecated - for soon-to-be removed features.
> 12. Removed - for now removed features.
> 13. Fixed - for any bug fixes.
> 14. Security - in case of vulnerabilities.

We are missing sections 11 and 12, just like 1 they are not really a common thing at all.
Also section 14 is neither reflected in your proposal nor in standing practise. Do we need it?

In any way, I suggest we put a template of these sections in changelog to sort the items into.
If you look at the file, the template sits at the end, but will be added on top for 4.12 etc.
This way, nobody has to come up with a section name by themselves, and bad habits are prevented.


@demonnic
> If we do it with the top section being 'PTB' for all changes in the upcoming release, and released versions listed beneath it, then the release notes could fairly easily be automated for release posts.

Indeed the top of the changelog will be the unreleased changes, which are present in PTB only.
Below that, the reverse chronological order lists previous releases.

I did not go back very far until now, so don't expect 1.0 in there. (Not even 4.0 presently)
May be worthwile to enhance the document to have all changes availanle in one place eventually.

Not sure which kind of automation you are hinting at here. May need to elaborate, please.

There are a few things done at release time still, that can't be prepared in changelog easily.

For example, we have a section "Credits" where all commiters are listed and specially thanked.
Maybe we could ask each person to add their name there if not present already for the release?

Another section like this follows with all translators named.
This surely can't be handled in git. May even drop it from template.

Finally we put a screenshot of git development statistics, which are to be gathered right before release.


@demonnic
> Slightly off-topic but related, it would also make it fairly easy to go through and bulk replace 'PTB' in the wiki as a mudlet version in the api docs to the version just released. This allows us to clearly mark functions which are only in the PTB without adding a lot of fiddly manual overhead.

This seems like it's possible already today without the file added to git at all?
Then again, the main problem is the wild growth of formulation for the version info.
It was cultivated a bit in the past, but not sure if that entangles PTB. But yes, another issue maybe.


@demonnic
> ... actually require things like changelog entries in order for PRs to pass ...
> I'm willing to work on setting this aspect of things up ...

As you say, some may fear increased strain on contributors so maybe postpone until we feel more confident. Then again, maybe we can settle on a friendly reminder "please add to changelog" not block merge for now?


@Edru2
> ... pretty bad at writing this kind of things ...

Thanks for voicing this point. We clearly must expect to review and offer help in writing that change item.

@Edru2
> Where exactly will this changelog be visible?

After merge, changelog will reside in each git branch like: 
https://github.com/Mudlet/Mudlet/blob/development/CHANGELOG.md
We may even link to this site from the README shown here:
https://github.com/Mudlet/Mudlet
If we like the file very much, maybe also from Mudlet website, or Mudlet app, etc.
Also of curse as part of Mudlet regular download and installation, in the Mudlet root directory on your disk drive.


@Delwing
> this file will be major source of git conflicts ...

I would not expect simply adding a line of text to be such a problem. We mostly don't modify lines, just add.
Thanks for linking all the examples though. I can now see how these issues may indeed become problematic.
For example, cherry-picking for a hotfix release would need attention to changelog entries added in between.
Those links also discuss the strategies and best practices those other projects finally settled upon.
I hope we can learn from those and include the necessary ones early. Need to review in more detail still.

@Delwing
> If we want to avoid conflicts instead of creating single file for each branch in changelog DIRECTORY ...
> .. Upon release all files should be merged automatically into single changelog ...

Do you have experience setting this up?

@SlySven
> .. should we sort by first word(s) in the texts or what? ...

I don't think this will be a necessary requirement to editing the changelog at all

@wiploo
> I don't think that a changelog.md is needed (I love changelos :-D)

Not sure how these two statements go together? :D

@wiploo
> When mudlet's releases are monthly based not many changes are listend in news page (10-20 rows). 

However still, somebody needs to go through all changes commits and write the change log.
Until now only few persons do this work in the last days before the release happens.
With changelog.md the work can be spread on more people and across time.
Also, you may want to review changelog for PTB like I said before.

@vadi2
> asking one person to wear the hat of a coder, a documentation writer, and a copywriter is a fair bit. 

On the other hand, we have peer review for PRs so people already collaborate and can help where necessary.
Some people indeed flourish writing text and are not as good at coding themselves, so may fit that role well.

@vadi2
> I realise that's ironic as I'm the one who does the release posts. 

We don't want you burned out of this work either and looking for ways to support.

@vadi2
> .. let's try this for a couple of months to see how people feel about it then.

Yes, the best way to judge the merits is actually putting it to practise and improve upon or remove it then.

Pondering the culprits and possible solutions with automation mentioned by @demonnic and @Delwing still.
We should start this in a smart way, but also not prepare solutions for months on end before touching ground.

Kebap committed:
Add sections for 4.12 development

Kebap committed:
Include 4.11 changelog

Kebap commented:
Marked as draft again to ponder the suggestions. 
Currently not enough bandwidth to push automatic solution.
Will not use in production before.

vadi2 commented:
Should we review this given the recent meeting and changelog-friendly PR titles?

Kebap commented:
Given the current PR-titles can be condensed into a handy changelog quite easily now (and are done so automatically each night for PTB information already) the work here is stopped for now. Can be revisited later if more interest arises again.
